pkg/controller/seed-controller-manager/ipam/controller.go

```diff
@@ -0,0 +1,272 @@
+		},
+	)
+	if err != nil {
+		log.Errorw("Reconciling failed", zap.Error(err))
+		r.recorder.Event(cluster, corev1.EventTypeWarning, "ReconcilingError", err.Error())
+	}
+	if result == nil {
+		result = &reconcile.Result{}
+	}
+	return *result, err
+}
+
+func (r *Reconciler) reconcile(ctx context.Context, cluster *kubermaticv1.Cluster) (*reconcile.Result, error) {
+	// List IPAM Pools
+	ipamPoolList := &kubermaticv1.IPAMPoolList{}
+	err := r.Client.List(ctx, ipamPoolList)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list IPAM pools: %w", err)
+	}
+
+	// Loop IPAM pools, considering only the relevant ones (i.e. IPAM pools with same cluster datacenter)
+	for _, ipamPool := range ipamPoolList.Items {
+		clusterDC := cluster.Spec.Cloud.DatacenterName
+
+		dcIPAMPoolCfg, isClusterDCConfigured := ipamPool.Spec.Datacenters[clusterDC]
+		if !isClusterDCConfigured {
+			// This IPAM pool is not relevant to cluster, so skip it
+			continue
+		}
+
+		ipamAllocation := &kubermaticv1.IPAMAllocation{}
+		err = r.Client.Get(ctx, types.NamespacedName{Namespace: fmt.Sprintf("cluster-%s", cluster.Name), Name: ipamPool.Name}, ipamAllocation)
+		if err == nil {
+			// skip because pool is already allocated for cluster
+			continue
+		} else if !apierrors.IsNotFound(err) {
+			return nil, err
+		}
+
+		dcIPAMPoolUsageMap, err := r.compileCurrentAllocationsForPoolInDatacenter(ctx, ipamPool.Name, dcIPAMPoolCfg)
+		if err != nil {
+			return nil, err
+		}
+
+		err = r.generateNewClusterAllocationForPool(ctx, cluster.Name, ipamPool.Name, dcIPAMPoolCfg, dcIPAMPoolUsageMap)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return nil, nil
+}
+
+func (r *Reconciler) compileCurrentAllocationsForPoolInDatacenter(ctx context.Context, ipamPoolName string, dcIPAMPoolCfg kubermaticv1.IPAMPoolDatacenterSettings) (datacenterIPAMPoolUsageMap, error) {
+	dcIPAMPoolUsageMap := newDatacenterIPAMPoolUsageMap()
+
+	// List all IPAM allocations
+	ipamAllocationList := &kubermaticv1.IPAMAllocationList{}
+	err := r.Client.List(ctx, ipamAllocationList)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list IPAM allocations: %w", err)
+	}
+
+	// Iterate current IPAM allocations to build a map of used IPs (for range allocation type)
+	// or used subnets (for prefix allocation type) per datacenter pool
+	for _, ipamAllocation := range ipamAllocationList.Items {
+		if ipamAllocation.Name != ipamPoolName {
+			// This allocation is not relevant for this IPAM Pool, so skip it
+			continue
+		}
+
+		switch ipamAllocation.Spec.Type {
+		case kubermaticv1.IPAMPoolAllocationTypeRange:
+			currentAllocatedIPs, err := getUsedIPsFromAddressRanges(ipamAllocation.Spec.Addresses)
+			if err != nil {
+				return nil, err
+			}
+			// check if the current allocation is compatible with the IPAMPool being applied
+			err = checkRangeAllocation(currentAllocatedIPs, string(dcIPAMPoolCfg.PoolCIDR), int(dcIPAMPoolCfg.AllocationRange))
+			if err != nil {
+				return nil, err
```

Comment:
I'm just thinking of potential causes of this kind of incompatibility. Change of IPAMPool? In any case, we should not forget to prevent such changes by webhook (may be part of follow-up PR).

Reply:
It's a way of allowing changes in IPAM Pool for some specific cases (i.e. if all allocations fit in the updated pool / e.g. just decreasing the pool prefix). Do you think it's better to block any IPAM Pool changes? Anyway, existing allocations won't be changed.

Reply:
Yeah, what I meant is to make sure we do not allow IPAMPool changes that would break this. But we there is an ticket for that already so does not have to be done in this PR: https://github.com/kubermatic/kubermatic/issues/9906